src/server/worker/datum_iterator.go

```diff
@@ -0,0 +1,295 @@
+	result.Reset()
+	return result, nil
+}
+
+func (d *pfsDatumIterator) Reset() {
+	d.location = -1
+}
+
+func (d *pfsDatumIterator) Len() int {
+	return len(d.inputs)
+}
+
+func (d *pfsDatumIterator) Datum() []*Input {
+	return []*Input{d.inputs[d.location]}
+}
+
+func (d *pfsDatumIterator) Next() bool {
+	d.location++
+	return d.location < len(d.inputs)
+}
+
+type unionDatumIterator struct {
+	inputs   []DatumIterator
+	unionIdx int
+}
+
+func newUnionDatumIterator(pachClient *client.APIClient, union []*pps.Input) (DatumIterator, error) {
+	result := &unionDatumIterator{}
+	for _, input := range union {
+		datumIterator, err := NewDatumIterator(pachClient, input)
+		if err != nil {
+			return nil, err
+		}
+		result.inputs = append(result.inputs, datumIterator)
+	}
+	result.Reset()
+	return result, nil
+}
+
+func (d *unionDatumIterator) Reset() {
+	for _, input := range d.inputs {
+		input.Reset()
+	}
+	d.unionIdx = 0
+}
+
+func (d *unionDatumIterator) Len() int {
+	result := 0
+	for _, datumIterator := range d.inputs {
+		result += datumIterator.Len()
+	}
+	return result
+}
+
+func (d *unionDatumIterator) Next() bool {
+	if len(d.inputs) == 0 {
+		return false
+	}
+	inputHasNext := d.inputs[d.unionIdx].Next()
+	if !inputHasNext {
+		d.unionIdx++
+		if d.unionIdx < len(d.inputs) {
+			return d.inputs[d.unionIdx].Next()
+		}
+	}
+	return false
+}
+
+func (d *unionDatumIterator) Datum() []*Input {
+	return d.inputs[d.unionIdx].Datum()
+}
+
+type crossDatumIterator struct {
+	inputs  []DatumIterator
+	started bool
+}
+
+func newCrossDatumIterator(pachClient *client.APIClient, cross []*pps.Input) (DatumIterator, error) {
+	result := &crossDatumIterator{}
+	for _, input := range cross {
+		datumIterator, err := NewDatumIterator(pachClient, input)
+		if err != nil {
+			return nil, err
+		}
+		result.inputs = append(result.inputs, datumIterator)
+	}
+	return result, nil
+}
+
+func (d *crossDatumIterator) Reset() {
+	for _, input := range d.inputs {
+		input.Reset()
+	}
+	d.started = false
+}
+
+func (d *crossDatumIterator) Len() int {
+	if len(d.inputs) == 0 {
+		return 0
+	}
+	result := d.inputs[0].Len()
+	for i := 1; i < len(d.inputs); i++ {
+		result *= d.inputs[i].Len()
+	}
+	return result
+}
+
+func (d *crossDatumIterator) Next() bool {
+	// we need to make sure each of the iterators has been started (i.e is not at -1)
+	hasNext := true
+	if !d.started {
+		for _, input := range d.inputs {
+			hasNext = hasNext && input.Next()
+		}
+		if hasNext {
+			d.started = true
+		}
+		return hasNext
+	}
+
+	// this is the normal case
+	for _, input := range d.inputs {
+		// if we're at the end of the "row"
+		if !input.Next() {
+			// we reset the "row"
+			input.Reset()
+
+			if !input.Next() {
+				// this should never happen, since the started case should handle this
+				return false
+			}
+			// after resetting this "row", start iterating through the next "row"
+		} else {
+			return true
+		}
+	}
+	return true
```

Comment:
This should be false right? When we get to the end (the last input gets reset), we are done right?